### PR TITLE
add Timetravel Effect

### LINKED ISSUE #3503: DoctorWho: Time Travel N Times
New Effect:

> For each suspended card you own and each permanent you control with a time counter on it, you may add or remove a time counter.

AI might need some logic on when it is better to add a time counter to a suspended card
normally it should do this:
* suspended Card: remove time counter
* Card in play: add time counter

The Effect probably should be designed like proliferate 

> To time travel, look at each permanent you control with a time counter on it and each card you own in exile with a time counter on it. For each of them, you choose whether you want to put a time counter on that card or permanent, remove a time counter from it, or do neither. Then those changes all happen simultaneously.

the rules say, you should check for each card if you want: Add/Remove/None, but imo its more clear if you select them first

### PR DESCRIPTION
**Pull request by Hanmac**
Closes #3503 

- [x] Add AI logic, use IsUsefulCounter
- [ ] Add Cards

i wanted to add the Tenth Doctor, but the AttackTrigger is a bit to tricky right now

### CODE REVIEW

**Comment by Agetian** [comment]
I can code the AI for this effect but I would need ay least one card for testing :)

**Comment by Hanmac** [comment]
@tool4ever 
>  Wibbly-wobbly, Timey-wimey {1}{U}
> Sorcery
> Time travel. (For each suspended card you own and each permanent you control with a time counter on it, you may add or remove a time counter.)
> Draw a card.

or 
> All of History, All at Once {2}{U}{U}
> Sorcery
> Time travel. (For each suspended card you own and each permanent you control with a time counter on it, you may add or remove a time counter.)
> Storm (When you cast this spell, copy it for each spell cast before it this turn.)

should be easy to script, the last one might need some AI logic for to not cast it when there are no cards with time counters in exile/battlefield?

**Comment by Agetian** [comment]
Ok, I'll give it a shot soon :)

**Comment by Agetian** [comment]
I added some simple AI, but it most likely will need some extending (for situations where you may want to add or remove a Time counter according to a different reasoning, or something else).

**Comment by Agetian** [comment]
Haven't found isUsefulCounter right off the bat, especially one that might refer to the Time counter, so feel free to extend this and refactor it into a method (isUsefulCounter) to determine when having a Time counter is a good thing or a bad thing.

File: forge-ai/src/main/java/forge/ai/ability/TimeTravelAi.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    @Override
+    public boolean chooseBinary(PlayerController.BinaryChoiceType kindOfChoice, SpellAbility sa, Map<String, Object> params) {
+        // Returning true means "add counter", false means "remove counter"
+
+        // TODO: extend this (usually, stuff in exile such as Suspended cards with Time counters is played once no Time counters are left,
+        // so removing them is good; stuff on the battlefield is usually stuff like Vanishing or As Foretold, which favors adding Time
+        // counters for better effect, but exceptions should be added here).
+        Card target = (Card)params.get("Target");
+        return !target.isInZone(ZoneType.Exile);
```
Comment (Hanmac) [inline forge-ai/src/main/java/forge/ai/ability/TimeTravelAi.java:39]: use `!ComputerUtil.isNegativeCounter`

then we have all at one place

**Comment by Hanmac** [comment]
there are some cards that might benefit from extra time counters in exile
like `Dinosaurs on a Spaceship` 

but for them we might add some more AI logic later

**Comment by Agetian** [comment]
It looks like lblChooseEntity is not yet added to the localization files, which causes a warning message and an incomplete prompt when choosing a card to add or remove counters from.

File: forge-ai/src/main/java/forge/ai/ability/TimeTravelAi.java (hunk 1 above)
Comment (Agetian) [reply forge-ai/src/main/java/forge/ai/ability/TimeTravelAi.java]: Done, good call! :)

**Review by Agetian — APPROVED** [review]

File: forge-ai/src/main/java/forge/ai/ability/TimeTravelAi.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+public class TimeTravelAi extends SpellAbilityAi {
+    @Override
+    protected boolean canPlayAI(Player aiPlayer, SpellAbility sa) {
+        boolean hasRelevantCards = false;
+
+        for (Card c : aiPlayer.getCardsIn(new ZoneType[] {ZoneType.Battlefield, ZoneType.Exile})) {
+            if (c.hasSuspend() || (c.isInPlay() && c.getCounters(CounterEnumType.TIME) >= 1)) {
+                hasRelevantCards = true;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Agetian) [reply forge-ai/src/main/java/forge/ai/ability/TimeTravelAi.java]: Done :)

**Review by Agetian — APPROVED** [review]

File: forge-ai/src/main/java/forge/ai/ability/TimeTravelAi.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import forge.game.spellability.SpellAbility;
+import forge.game.zone.ZoneType;
+
+import java.util.Map;
+
+public class TimeTravelAi extends SpellAbilityAi {
+    @Override
+    protected boolean canPlayAI(Player aiPlayer, SpellAbility sa) {
+        boolean hasSuspendedCards = !CardLists.filter(aiPlayer.getCardsIn(ZoneType.Exile), CardPredicates.hasSuspend()).isEmpty();
+        boolean hasRelevantCardsOTB = !CardLists.filter(aiPlayer.getCardsIn(ZoneType.Battlefield), CardPredicates.hasCounter(CounterEnumType.TIME)).isEmpty();
```
Comment (Hanmac) [inline forge-ai/src/main/java/forge/ai/ability/TimeTravelAi.java:20]: `Iterables.any` probably better `CardLists.filter`

**Review by tool4ever — APPROVED** [review]

**Review by Agetian — APPROVED** [review]

**Comment by Agetian** [comment]
Yeah, feel free to extend the logic a bit, sounds like good additions :)

**Comment by Hanmac** [comment]
@tool4ever is that good enough for you to start with the cards?

**Comment by tool4ever** [comment]
My main focus isn't really new scripts but if it needs further tweaks I'll be around to support it

**Comment by Hanmac** [comment]
@tehdiplomat is this okay to be merged, or should we add more cards?

**Pull request merged**